### PR TITLE
MM-15445 Test for the user's profile pop-over (#10806)

### PR DESCRIPTION
**Pull request by VolatianaYuliana**
Write an automated test using Cypress for "Join/leave messages (including combined): Usernames are links, open profile pop-overs"

[Jira ticket
](https://mattermost.atlassian.net/browse/MM-15445)

Fixes https://github.com/mattermost/mattermost-server/issues/10806

### CODE REVIEW

File: components/multiselect/multiselect_list.jsx [hunk 1]
```diff
@@ -160,6 +160,7 @@ export default class MultiSelectList extends React.Component {
             <div className='more-modal__list'>
                 <div
                     ref='list'
+                    id='userAddList'
```
Comment (saturninoabril) [inline components/multiselect/multiselect_list.jsx]: This component is generic, not only for user selection.  I recommend to change this, maybe something like: `multiSelectList`

File: cypress/integration/channel/add_users_to_channel_spec.js [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+/*eslint max-nested-callbacks: ["error", 5]*/
+
+describe('Add user to a channel', () => {
+    it('CS15445 Join/leave messages (including combined): Usernames are links, open profile popovers', () => {
+        // # Login as 'user-1' and go to /
+        cy.apiLogin('user-1');
+        cy.visit('/');
+
+        // # Select 'doloremque' channel on the left hand side
+        cy.get('#sidebarItem_saepe-5').click();
```
Comment (saturninoabril) [inline cypress/integration/channel/add_users_to_channel_spec.js]: I'm getting test failure when using `saepe-5` and `doloremque`.  Instead of using those channels, I suggests to create 2 test channels (see `cy.apiCreateChannel`).

File: cypress/integration/channel/add_users_to_channel_spec.js [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+        cy.get('#channelHeaderDropdownMenu').should('be.visible');
+
+        // # Click 'Add Members'
+        cy.get('#channelAddMembers').click();
+
+        // * Assert that modal appears
+        cy.get('#addUsersToChannelModal').should('be.visible');
+
+        // # Wait 1 second
+        cy.wait(1000);// eslint-disable-line
```
Comment (saturninoabril) [inline cypress/integration/channel/add_users_to_channel_spec.js]: If wait is necessary, please use `TIMEOUTS` constant, probably TINY.

File: cypress/integration/channel/add_users_to_channel_spec.js [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+        // # Click the first row clickable of the modal to select a user
+        cy.get('#userAddList').first().click();
+
+        // # Click the button "Add" to add this user to 'doloremque' channel
+        cy.get('#saveItems').click();
+
+        cy.getLastPostId().then((id) => {
+            // # On the system message advising a user was added to the channel, click the username link
```
Comment (saturninoabril) [inline cypress/integration/channel/add_users_to_channel_spec.js]: I suggest to partially verify text content of system message, like:
``cy.get(`#postMessageText_${id}`).should('contain', 'added to the channel by you');``

While to other instance as,
``cy.get(`#postMessageText_${id}`).should('contain', '2 others were added to the channel by you');``

File: cypress/integration/channel/add_users_to_channel_spec.js [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+            cy.get(`#post_${id}`).find('.mention-link').first().click({force: true}).then(($el) => {
+                // * Then the user's profile pop-over should be displayed when their username is clicked, and the pop-over is fully visible
+                cy.get('#user-profile-popover').should('be.visible');
+
+                // * The pop-over title  at the top of the pop-over should be the same as the username link
+                const userName = $el[0].innerHTML;
+                const escapedUsername = userName.replace('@', '').replace('.', '-');
+                cy.get(`#userPopoverUserName-${escapedUsername}`).should('contain', userName);
+            });
```
Comment (saturninoabril) [inline cypress/integration/channel/add_users_to_channel_spec.js]: This has almost the same pattern as the other checks.  I suggest to refactor it as separate function, like:
```javascript
function verifyMentionedUserAndProfilePopover(postId) {
    cy.get(`#post_${postId}`).find('.markdown__paragraph-inline').find('.mention-link').each(($el) => {
        // # Click each username link
        cy.wrap($el).click();

        // * Should display user's profile pop-over
        const userName = $el[0].innerHTML;
        cy.get('#user-profile-popover').should('be.visible');

        // * Pop-over title should be the same as the username link
        const escapedUsername = userName.replace('@', '').replace('.', '-');
        cy.get(`#userPopoverUserName-${escapedUsername}`).should('contain', userName);
    });
}
```

File: cypress/integration/channel/add_users_to_channel_spec.js [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+
+                    // * The profile pop-over of each user should be displayed when their username is clicked, and the pop-over is fully visible
+                    const userName = $el[0].innerHTML;
+                    cy.get('#user-profile-popover').should('be.visible');
+
+                    // * The pop-over title  at the top of the pop-over should be the same as the username link for each user
+                    const escapedUsername = userName.replace('@', '').replace('.', '-');
+                    cy.get(`#userPopoverUserName-${escapedUsername}`).should('contain', userName);
+                });
+            });
```
Comment (saturninoabril) [inline cypress/integration/channel/add_users_to_channel_spec.js]: If above is applied, then this can be simplified as,
```javascript
// # System message should contain message that users were added to the channel
cy.get(`#postMessageText_${id}`).should('contain', '2 others were added to the channel by you');

// # Click "2 others" to expand more users
cy.get(`#post_${id}`).find('.markdown__paragraph-inline').siblings('a').first().click();

// * Verify each username link
verifyMentionedUserAndProfilePopover(id);
```

File: cypress/integration/channel/autocomplete_shown_each_channel_spec.js [hunk 7]
```diff
@@ -3,7 +3,7 @@
 
 // ***************************************************************
 // - [#] indicates a test step (e.g. 1. Go to a page)
-// - [*] indicates an assertion (e.g. * Check the title)
+// - [*] in dicates an assertion (e.g. * Check the title)
```
Comment (saturninoabril) [inline cypress/integration/channel/autocomplete_shown_each_channel_spec.js]: Probably unintentional change; please revert.

File: cypress/integration/channel/add_users_to_channel_spec.js [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+        // # Click 3 times the first row clickable of the modal to select 3 different users
+        cy.get('#userAddList').first().click();
+        cy.get('#userAddList').first().click();
+        cy.get('#userAddList').first().click();
+
+        // # Click the button "Add" to add these 3 users to 'sint' channel
+        cy.get('#saveItems').click();
+
+        // # Wait 2 seconds to allow the popup to disappear
+        cy.wait(2000); // eslint-disable-line
```
Comment (thekiiingbob) [inline cypress/integration/channel/add_users_to_channel_spec.js]: `cy.get('#foobarbaz').should('be.hidden')` or cy.get('#foobarbaz').should('not.exist')` (depending on what happens to it) for the appropriate element instead of waiting.

**Comment by VolatianaYuliana** [comment]
Hi! thanks for the reviews, I'm gonna check all those comments out :)

File: cypress/integration/channel/add_users_to_channel_spec.js (hunk 8 above)
Comment (VolatianaYuliana) [reply cypress/integration/channel/add_users_to_channel_spec.js]: Hey @thekiiingbob , I tried but it didn't work so I used TIMEOUT instead

**Comment by VolatianaYuliana** [comment]
Hi @saturninoabril ! I fixed it

**Comment by saturninoabril** [comment]
@VolatianaYuliana Thanks, I'll look into it again, probably later today

File: cypress/integration/channel/add_users_to_channel_spec.js (hunk 8 above)
Comment (thekiiingbob) [reply cypress/integration/channel/add_users_to_channel_spec.js]: @VolatianaYuliana `cy.get('#addUsersToChannelModal').should('not.exist');` for the 2 places you were waiting for the modal to go away, worked for me. Was there a specific error/issue you were running into?

File: cypress/integration/channel/add_users_to_channel_spec.js [hunk 9]
```diff
@@ -1,101 +1,136 @@
...
-        // * The dropdown menu of te channel header should be visible;
-        cy.get('#channelHeaderDropdownMenu').should('be.visible');
+                // # Click 'Add Members'
+                cy.get('#channelAddMembers').click();
 
-        // # Click 'Add Members'
-        cy.get('#channelAddMembers').click();
+                // * Assert that modal appears
+                cy.get('#addUsersToChannelModal').should('be.visible');
+                cy.wait(TIMEOUTS.TINY);
```
Comment (thekiiingbob) [inline cypress/integration/channel/add_users_to_channel_spec.js]: @VolatianaYuliana Whats the reasoning for this wait? Modal should be visible at this point, was `#multiSelectList` not visible? If so, tacking on an assertion to `cy.get('#multiSelectList')` should suffice.

File: cypress/integration/channel/add_users_to_channel_spec.js (hunk 8 above)
Comment (VolatianaYuliana) [reply cypress/integration/channel/add_users_to_channel_spec.js]: This is what I got when I use cy.get('#addUsersToChannelModal').should('not.exist') instead of waiting... but it works when I use TIMEOUT.
error

File: cypress/integration/channel/add_users_to_channel_spec.js (hunk 9 above)
Comment (VolatianaYuliana) [reply cypress/integration/channel/add_users_to_channel_spec.js]: You're right. I'm gonna get rid of it @thekiiingbob

File: cypress/integration/channel/add_users_to_channel_spec.js (hunk 8 above)
Comment (thekiiingbob) [reply cypress/integration/channel/add_users_to_channel_spec.js]: @VolatianaYuliana Looks like the wait for the element to not exists works properly, but the post isn't ready yet? Locally, even with induced network latency, I cannot reproduce this error. Is the post still loading for you? Seems weird because the check for the post to have and id and also not be pending looks to be passing for you so it SHOULD be ready.

Comment (VolatianaYuliana) [reply cypress/integration/channel/add_users_to_channel_spec.js]: The 'added to the channel by you' text is not yet visible in the last post even if the check for the last post to have id is passing  ...as you can see in the screenshot, 
error2

Comment (VolatianaYuliana) [reply cypress/integration/channel/add_users_to_channel_spec.js]: it's weird why it doesn't work for me

Comment (thekiiingbob) [reply cypress/integration/channel/add_users_to_channel_spec.js]: Mind if I get the specs of your development machine? It feels like it's a timing issue of the system not posting the new message for you as quick as we want it to. Ideally we don't want to do a hard wait, so let me think about what we can wait for.

Comment (VolatianaYuliana) [reply cypress/integration/channel/add_users_to_channel_spec.js]: Here are my machine specs: 
specs

**Comment by VolatianaYuliana** [comment]
Hey @thekiiingbob , I reorganized the test and I'm still waiting for your reply about how we can wait for instead of "TIMEOUTS".

**Comment by thekiiingbob** [comment]
> Hey @thekiiingbob , I reorganized the test and I'm still waiting for your reply about how we can wait for instead of "TIMEOUTS".

@VolatianaYuliana I was able to get that same error removing the waits and using Fast 3G in the network settings in chrome. I added the `cy.get('#addUsersToChannelModal').should('not.exist')` right after both of the `cy.get('#saveItems').click();` lines and I cannot get the same error now. Can you reconfirm? Merging in master may have addressed some of the issues.

**Comment by VolatianaYuliana** [comment]
@thekiiingbob it still doesn't work on my side but I guess that we can use `cy.get('#addUsersToChannelModal').should('not.exist')` anyway if you say it works as the issue may be related with my config.

**Comment by VolatianaYuliana** [comment]
@saturninoabril  @thekiiingbob  I fixed it

File: cypress/integration/channel/add_users_to_channel_spec.js [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+                cy.get('#channelHeaderDropdownMenu').should('be.visible');
+
+                // # Click 'Add Members'
+                cy.get('#channelAddMembers').click();
+
+                // * Assert that the modal appears
+                cy.get('#addUsersToChannelModal').should('be.visible');
+
+                // # Click 3 times the first row clickable of the modal to select 3 different users
+                cy.get('#multiSelectList').first().click();
```
Comment (thekiiingbob) [inline cypress/integration/channel/add_users_to_channel_spec.js]: @VolatianaYuliana I think starting here is where the tests start to differ, but before here they share a lot of the same code. Can you rollup the shared "setup" stuff into a function? Maybe `addUsersToTeam(amount)`? And that function could click on `cy.get('#multiSelectList').first().click();` however many times you need it to?

**Comment by VolatianaYuliana** [comment]
I updated the code @thekiiingbob

File: cypress/integration/channel/add_users_to_channel_spec.js [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+    it('Single User: Usernames are links, open profile popovers', () => {
+        // # Login as 'user-1', go to / and create 'Channel test 1'
+        cy.apiLogin('user-1');
+        cy.visit('/');
+        cy.getCurrentTeamId().then((teamId) => {
+            cy.apiCreateChannel(teamId, 'channel-test-1', channelDisplayNameTest1).then((res) => {
+                channelTest1Id = res.body.name;
+
+                // # Select 'Channel test 1' channel on the left hand side
+                cy.get(`#sidebarItem_${channelTest1Id}`).click();
```
Comment (saturninoabril) [inline cypress/integration/channel/add_users_to_channel_spec.js]: This is unable to click if there are too many channels in the sidebar.  So I recommend,
```suggestion
                cy.get(`#sidebarItem_${channelTest1Id}`).scrollIntoView().click({force: true});
```

File: cypress/integration/channel/add_users_to_channel_spec.js [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+    });
+
+    it('Combined Users: Usernames are links, open profile popovers', () => {
+        cy.getCurrentTeamId().then((teamId) => {
+            // # Create 'Channel test 2'
+            cy.apiCreateChannel(teamId, 'channel-test-2', channelDisplayNameTest2).then((res) => {
+                channelTest2Id = res.body.name;
+
+                // # Select 'Channel test 2' channel on the left hand side
+                cy.get(`#sidebarItem_${channelTest2Id}`).click();
```
Comment (saturninoabril) [inline cypress/integration/channel/add_users_to_channel_spec.js]: Same here.
```suggestion
                cy.get(`#sidebarItem_${channelTest2Id}`).scrollIntoView().click({force: true});
```

File: cypress/integration/channel/add_users_to_channel_spec.js [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+        cy.apiLogin('user-1');
+        cy.visit('/');
+        cy.getCurrentTeamId().then((teamId) => {
+            cy.apiCreateChannel(teamId, 'channel-test-1', channelDisplayNameTest1).then((res) => {
+                channelTest1Id = res.body.name;
+
+                // # Select 'Channel test 1' channel on the left hand side
+                cy.get(`#sidebarItem_${channelTest1Id}`).click();
+
+                // * 'Channel test 1' shoul be visible at the top of the center pane
```
Comment (saturninoabril) [inline cypress/integration/channel/add_users_to_channel_spec.js]: Spell check
```suggestion
                // * 'Channel test 1' should be visible at the top of the center pane
```

**Comment by VolatianaYuliana** [comment]
Hey @saturnoabril.okay I'll change that 😊

**Comment by jasonblais** [comment]
Hey @VolatianaYuliana, just touching base if you had any questions on the latest feedback? :)

**Comment by VolatianaYuliana** [comment]
Hi! Sorry I lost track of this issue, I'm gonna push uptades by the end of this week :)

File: cypress/integration/channel/add_users_to_channel_spec.js [hunk 14]
```diff
@@ -0,0 +1,140 @@
...
+        cy.get('#user-profile-popover').should('be.visible');
+
+        // * The pop-over title  at the top of the pop-over should be the same as the username link for each user
+        const escapedUsername = userName.replace('@', '').replace('.', '-');
+        cy.get(`#userPopoverUserName-${escapedUsername}`).should('contain', userName);
+    });
+}
+
+function addUsersToChannel(amount) {
+    for (let i = 0; i < amount; i++) {
```
[the remark this answers is not in the document]
Comment (VolatianaYuliana) [reply cypress/integration/channel/add_users_to_channel_spec.js]: @thekiiingbob fixed

**Comment by VolatianaYuliana** [comment]
Hey @saturninoabril, this is now up to date.

File: components/channel_invite_modal/channel_invite_modal.jsx [hunk 15]
```diff
@@ -159,6 +159,7 @@ export default class ChannelInviteModal extends React.Component {
 
         return (
             <div
+                id={`modalUserRow-${option.id}`}
```
Comment (saturninoabril) [inline components/channel_invite_modal/channel_invite_modal.jsx]: This ID is not used elsewhere.

File: components/profile_popover/profile_popover.jsx [hunk 16]
```diff
@@ -440,8 +440,10 @@ class ProfilePopover extends React.PureComponent {
         if (this.props.hasMention) {
             title = <a onClick={this.handleMentionKeyClick}>{title}</a>;
         }
+
+        const escapedUsername = this.props.user.username.replace('.', '-');
         title = (
-            <span>
+            <span id={`userPopoverUserName-${escapedUsername}`}>
```
Comment (saturninoabril) [inline components/profile_popover/profile_popover.jsx]: Per recent agreement, we'll be using `data-testid` especially for this case where dynamic ID can be conformant HTML id attribute

File: e2e/cypress/fixtures/timeouts.js [hunk 17]
```diff
@@ -3,6 +3,7 @@
 
 module.exports = {
     TINY: 500,
+    SHORT: 3000,
```
Comment (saturninoabril) [inline e2e/cypress/fixtures/timeouts.js]: This is not used elsewhere.

File: e2e/cypress/integration/channel/autocomplete_shown_each_channel_spec.js [hunk 18]
```diff
@@ -3,7 +3,7 @@
 
 // ***************************************************************
 // - [#] indicates a test step (e.g. 1. Go to a page)
-// - [*] indicates an assertion (e.g. * Check the title)
+// - [*] in dicates an assertion (e.g. * Check the title)
```
Comment (saturninoabril) [inline e2e/cypress/integration/channel/autocomplete_shown_each_channel_spec.js]: Should revert this change.

File: components/profile_popover/profile_popover.test.jsx [hunk 19]
```diff
@@ -43,6 +43,7 @@ describe('components/ProfilePopover', () => {
             user: {
                 is_bot: true,
                 bot_description: 'bot description',
+                username: 'john.doe',
```
Comment (saturninoabril) [inline components/profile_popover/profile_popover.test.jsx]: Not related to this PR.
